### PR TITLE
fix: support Jackson 3 based JsonMapper in CamundaAssertJsonMapper

### LINKED ISSUE #60484: CamundaAssertJsonMapper hardcodes Jackson 2 JsonNode, breaks hasVariable()/hasVariables() when the client's JsonMapper is Jackson3JsonMapperConfiguration's Jackson-3-based CamundaJackson3ObjectMapper
### Affected version
8.9.16 (also reproduces against the current `main` branch source of `CamundaAssertJsonMapper.java` - unreleased, so 8.10.0 will have this too)

### Severity
Medium

### Likelihood
High - affects any Spring Boot 4.x project (Jackson 3 by default) using `camunda-spring-boot-starter` + `camunda-process-test-spring`, as soon as a test asserts a variable value via `hasVariable(name, value)`.

### Description

`camunda-spring-boot-starter`'s `Jackson3JsonMapperConfiguration` auto-configures `io.camunda.client.impl.CamundaJackson3ObjectMapper` (backed by `tools.jackson.databind.ObjectMapper`, Jackson 3) as the `CamundaClient`'s default `JsonMapper` whenever Jackson 3 is on the classpath - which it is by default under Spring Boot 4.x (`spring-boot-starter-jackson` / `spring-boot-starter-webflux` pull in `tools.jackson.core:jackson-databind`).

`camunda-process-test-java`'s `CamundaAssertJsonMapper` (used internally by `CamundaAssert.hasVariable()`/`hasVariables()`) still hardcodes `com.fasterxml.jackson.databind.JsonNode` (classic Jackson 2, an abstract class) as its universal JSON-tree representation:

```java
// testing/camunda-process-test-java/.../util/CamundaAssertJsonMapper.java (current main)
import com.fasterxml.jackson.databind.JsonNode;
...
private JsonNode write(final Object value) {
  return jsonMapper.fromJson(jsonMapper.toJson(value), JsonNode.class);
}
```

When `jsonMapper` is the Jackson-3-based `CamundaJackson3ObjectMapper`, asking it to deserialize into the Jackson-2 abstract class `com.fasterxml.jackson.databind.JsonNode` fails outright - Jackson 3 has no concept of that type at all, so it can't even fall back to a compatible concrete subtype.

Note this is a different, more fundamental bug than #55612 / PR #55827 ("normalize numeric JsonNode types in CPT variable assertions") - that fix was about `JsonNode` sub-type mismatches (`LongNode` vs `IntNode` etc.) when both sides are already Jackson 2. This one reproduces for a plain `Boolean`, and is about the *JsonMapper implementation itself* being from a different Jackson generation than the hardcoded target type.

### Steps to reproduce

```java
@SpringBootTest(properties = {
    // any Spring Boot 4.x project with Jackson 3 on the classpath (spring-boot-starter-webflux
    // or spring-boot-starter-jackson) reproduces this with camunda-spring-boot-starter's default
    // JsonMapper wiring - no special config needed to select Jackson3JsonMapperConfiguration.
})
@CamundaSpringProcessTest
class JacksonMismatchIT {

    @Autowired
    private CamundaClient client;

    @Test
    void hasVariableFailsForAPlainBoolean() {
        client.newDeployResourceCommand()
                .addProcessModel(
                        Bpmn.createExecutableProcess("process").startEvent().endEvent().done(),
                        "process.bpmn")
                .send().join();

        var pi = client.newCreateInstanceCommand()
                .bpmnProcessId("process")
                .latestVersion()
                .variables(Map.of("flag", true))
                .send().join();

        CamundaAssert.assertThat(pi).hasVariable("flag", true); // fails
    }
}
```

### Current behavior

```
io.camunda.process.test.impl.assertions.util.CamundaAssertJsonMapper$JsonMappingException: Failed to transform value to JSON: 'true'
	at io.camunda.process.test.impl.assertions.util.CamundaAssertJsonMapper.toJsonNode(CamundaAssertJsonMapper.java:63)
	at io.camunda.process.test.impl.assertions.VariableAssertj.hasVariable(VariableAssertj.java:146)
	at io.camunda.process.test.impl.assertions.VariableAssertj.hasVariable(VariableAssertj.java:136)
	at io.camunda.process.test.impl.assertions.ProcessInstanceAssertj.hasVariable(ProcessInstanceAssertj.java:292)
Caused by: io.camunda.client.api.command.InternalClientException: Failed to deserialize json 'true' to class 'class com.fasterxml.jackson.databind.JsonNode'
	at io.camunda.client.impl.CamundaJackson3ObjectMapper.fromJson(CamundaJackson3ObjectMapper.java:57)
	at io.camunda.process.test.impl.assertions.util.CamundaAssertJsonMapper.write(CamundaAssertJsonMapper.java:81)
	at io.camunda.process.test.impl.assertions.util.CamundaAssertJsonMapper.toJsonNode(CamundaAssertJsonMapper.java:60)
Caused by: tools.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `com.fasterxml.jackson.databind.JsonNode` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information
```

I hit this via `CamundaAssert.assertThat(processInstanceSelector).hasVariable("isRefund", true)` - explicitly setting `CamundaAssert.setJsonMapper(client.getConfiguration().getJsonMapper())` makes no difference, since that IS the Jackson-3-based mapper causing the failure. Workaround: use `hasVariableSatisfies(name, Boolean.class, assertion)` instead, which reads the actual value directly and doesn't round-trip through `JsonNode`.

### Expected behavior

`hasVariable()`/`hasVariables()` should work regardless of whether the `CamundaClient`'s configured `JsonMapper` is Jackson-2- or Jackson-3-based (e.g. by having `CamundaAssertJsonMapper` use the `JsonMapper`'s own tree-node type / `Map`/`Object` comparison instead of a hardcoded `com.fasterxml.jackson.databind.JsonNode` target type).

### Additional context

- All dependencies confirmed pinned to the same `8.9.16` when reproducing (`camunda-client-java`, `camunda-spring-boot-starter`, `camunda-process-test-spring`, `camunda-process-test-java`) - this isn't a version-skew issue between Camunda artifacts.
- This isn't an unsupported combination on the consumer's end either: `camunda-process-test-spring:8.9.16`'s own POM declares `spring-boot 4.0.7` as its native dependency, and so does `camunda-spring-boot-starter:8.9.16`'s POM. Spring Boot 4.x defaults to Jackson 3, so `Jackson3JsonMapperConfiguration` activating is the intended, officially-supported default for this release - not an edge case triggered only by an unusual dependency override.
- Confirmed still present by reading `testing/camunda-process-test-java/src/main/java/io/camunda/process/test/impl/assertions/util/CamundaAssertJsonMapper.java` directly off `main` - unreleased, so this isn't fixed in the upcoming 8.10 line either.
- Originates from #46549 ("Missing Jackson 3 support for Spring Boot 4"), which listed "allowing camunda-process-test assertions based on Jackson 3" as in-scope and was closed as completed, but `CamundaAssertJsonMapper` doesn't appear to have been updated as part of that work.

### PR DESCRIPTION
**Pull request by mustafadagher**
## Description

`CamundaAssertJsonMapper` (used internally by `CamundaAssert.hasVariable()`/`hasVariables()`) hardcoded `com.fasterxml.jackson.databind.JsonNode` (Jackson 2) as the target type when asking the client's `JsonMapper` to deserialize JSON. When the `CamundaClient` is configured with the Jackson 3 based `CamundaJackson3ObjectMapper` (the default on Spring Boot 4.x via `camunda-spring-boot-starter`'s `Jackson3JsonMapperConfiguration`), this failed outright since Jackson 3 has no concept of the Jackson 2 `JsonNode` abstract type.

This PR removes the dependency on any JSON-library-specific tree type altogether, rather than working around it by building JSON trees with a separate, internally-owned Jackson 2 mapper:

- `CamundaAssertJsonMapper` now represents JSON values as plain Java types (`Map`, `List`, `String`, `Number`, `Boolean`, `null`) instead of a `JsonNode`. Every conversion — both reading actual values and converting expected values — goes entirely through the client's own `JsonMapper` (`fromJson(value, Object.class)` and `toJson(value)`), so a `JsonMapper` implementation only needs to honor its existing contract; no tree-building capability is assumed.
- Added `toJsonValue(Object)` (replacing `toJsonNode`), which serializes an expected value to JSON and reads it back as a plain `Object` so it can be structurally compared against actual values produced by `readJson`. Numeric/type normalization (int vs long, float vs double, etc.) still happens this way.
- `readJson(String)` now returns `Object` instead of `JsonNode`; `toJson(Object)` renders a value back to a JSON string for building assertion failure messages (e.g. expected/actual variable dumps).
- Updated callers (`VariableAssertj`, `DecisionOutputAssertj`) and `DocumentReferenceResolver` to walk plain `Map`/`List` structures instead of `JsonNode` trees when resolving Camunda document references, including tolerating JSON `null` values encountered as Java `null`.
- Added `Jackson3JsonMapper`, a `JsonMapper` test double built on Jackson 3 (`tools.jackson.*`) that cannot produce Jackson 2 trees, to prove the assertions rely only on the `JsonMapper` contract. Used it in a new `JsonMapperAssertTest` covering both `hasVariable`/`hasVariables` (expected-value path) and `hasOutput` (actual-value path), since each fails independently if only one conversion direction is fixed.
- Added `Jackson3JsonMapperIT` in `camunda-process-test-spring`, an end-to-end Spring Boot test configuring a Jackson 3 `ObjectMapper` bean so `Jackson3JsonMapperConfiguration` wires in `CamundaJackson3ObjectMapper`, verifying variable assertions work when the client is actually configured this way. This test is excluded from the `camunda-process-test-spring-boot-3` module's recompiled `src/test/java` (via `testExcludes` on `maven-compiler-plugin`), since that module only has Jackson 2 on its classpath.
- Added `jackson-databind` (Jackson 3, `tools.jackson.core`) as a test-scoped dependency in `camunda-process-test-java` and `camunda-process-test-spring`.

## Checklist

- [x] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes), [for CI changes](https://camunda.github.io/camunda/ci/#when-to-backport-ci-changes), or [for documentation changes](https://camunda.github.io/camunda/ci/#documentation-specific-backporting-monorepo-docs-folders)).
- [ ] If this PR modifies the [C8 Orchestration Cluster E2E Test Suite](qa/c8-orchestration-cluster-e2e-test-suite), the relevant tests have been run via the [on-demand workflow](https://github.com/camunda/camunda/actions/workflows/c8-orchestration-cluster-e2e-tests-on-demand.yml) before requesting review. Any test failures are documented in the PR description and confirmed not to be regressions introduced by this PR.

## Related issues

closes #60484

### CODE REVIEW

File: testing/camunda-process-test-java/src/main/java/io/camunda/process/test/impl/assertions/util/CamundaAssertJsonMapper.java [hunk 1]
```diff
@@ -15,21 +15,48 @@
...
 
 public class CamundaAssertJsonMapper {
 
+  /**
+   * Builds the JSON trees that the assertions compare. The client's {@link JsonMapper} cannot be
+   * asked for them: its contract has no notion of {@link JsonNode}, so implementations that are not
+   * based on Jackson 2 cannot return one. Converting between objects and JSON strings still goes
+   * through the client's mapper, so custom serializers keep applying.
+   */
+  private static final ObjectMapper JSON_TREE_MAPPER = new ObjectMapper();
```
Comment (saig0) [inline testing/camunda-process-test-java/src/main/java/io/camunda/process/test/impl/assertions/util/CamundaAssertJsonMapper.java]: 💭 Can we get rid of the JSON tree mapper if we replace `JsonNode` with `Object` or use a different approach? I prefer to use only one object mapper instead of two.

File: testing/camunda-process-test-spring/src/test/jackson3/java/io/camunda/process/test/api/Jackson3AssertionsTest.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+ */
+@ExtendWith(MockitoExtension.class)
+public class Jackson3AssertionsTest {
+
+  private static final long PROCESS_INSTANCE_KEY = 1L;
+  private static final String DECISION_NAME = "decision";
+  private static final String DECISION_INSTANCE_ID = "decision-instance-id";
+
+  /** Keeps assertions that are expected to fail from waiting for the default timeout. */
+  private static final Duration EXPECT_FAILURE_TIMEOUT = Duration.ofMillis(200);
```
Comment (saig0) [inline testing/camunda-process-test-spring/src/test/jackson3/java/io/camunda/process/test/api/Jackson3AssertionsTest.java]: 🔧 We should use `@CamundaAssertExpectFailure` instead of modifying the assertion timeout. See https://github.com/camunda/camunda/blob/main/testing/CONTRIBUTING.md#4-testing-conventions.

File: testing/camunda-process-test-spring/pom.xml [hunk 3]
```diff
@@ -201,6 +210,31 @@
         <artifactId>revapi-maven-plugin</artifactId>
       </plugin>
 
+      <!--
+        camunda-process-test-spring-boot-3 has no test sources of its own and re-compiles
+        src/test/java against the Spring Boot 3 stack, which carries Jackson 2 only. Jackson 3 is
+        requested as a test dependency here, so anything using it has to stay out of that shared
+        source root or it fails there with NoClassDefFoundError.
+      -->
```
Comment (saig0) [inline testing/camunda-process-test-spring/pom.xml]: 💭 Instead of placing the tests in a different source and adding them via a plugin, we could add them to the regular package and exclude them in the `camunda-process-test-spring-boot-3` module. This way, we keep the default module clean.

File: testing/camunda-process-test-java/src/main/java/io/camunda/process/test/impl/assertions/util/CamundaAssertJsonMapper.java [hunk 4]
```diff
@@ -15,21 +15,48 @@
...
 
   public JsonNode readJson(final String value) {
-    return readJson(value, JsonNode.class, NullNode.getInstance());
+    if (value == null) {
+      return NullNode.getInstance();
+    }
+
+    final JsonNode jsonNode;
+    try {
+      jsonNode = JSON_TREE_MAPPER.readTree(value);
```
Comment (saig0) [inline testing/camunda-process-test-java/src/main/java/io/camunda/process/test/impl/assertions/util/CamundaAssertJsonMapper.java]: 💭 I'm wondering if we would run into the issue https://github.com/camunda/camunda/issues/37758 again if you use the default JSON mapper instead of the custom one from the client. The concrete issue is solved. However, the default value mapper might behave differently from the custom value mapper in some cases.

**Review by saig0 — CHANGES_REQUESTED** [review]
@mustafadagher looks good. 👍 

🔧 Minor: align test style.

💭 Some concerns about the approach of using another object mapper.

File: testing/camunda-process-test-spring/pom.xml (hunk 3 above)
Comment (mustafadagher) [reply testing/camunda-process-test-spring/pom.xml]: I had it this way originally, but `tools.jackson.databind.ObjectMapper` kept showing red in intellij (compilation error). Apparently, intellij ignores `maven-compiler-plugin`'s `testExcludes`.

Comment (saig0) [reply testing/camunda-process-test-spring/pom.xml]: Sounds weird. :sweat_smile: However, it is not a big issue.

Comment (saig0) [reply testing/camunda-process-test-spring/pom.xml]: @mustafadagher please take a look at https://github.com/camunda/camunda/blob/main/clients/camunda-spring-boot-3-starter/pom.xml#L139-L156. There, we use the pattern to exclude Jackson 3 tests for the Camunda Spring Boot Starter 3.

File: testing/camunda-process-test-java/src/main/java/io/camunda/process/test/impl/judge/DocumentReferenceResolver.java [hunk 5]
```diff
@@ -82,59 +81,56 @@ public List<ResolvedDocument> resolve(final String variableJson) {
...
-      node.forEach(element -> collectReferences(element, out));
-    } else if (node.isObject()) {
-      node.fieldNames().forEachRemaining(fieldName -> collectReferences(node.get(fieldName), out));
+    @SuppressWarnings("unchecked")
+    final Map<String, Object> objectNode = (Map<String, Object>) node;
+    if (isDocumentReference(objectNode)) {
+      out.add(objectNode);
+      return;
     }
+    objectNode.values().forEach(value -> collectReferences(value, out));
```
[the remark this answers is not in the document]
Comment (mustafadagher) [reply testing/camunda-process-test-java/src/main/java/io/camunda/process/test/impl/judge/DocumentReferenceResolver.java:121]: Not a behaviour change, main has the identical early return:
```java
if (isDocumentReference(node)) {
  out.add(node);
  return;
}
```
The javadoc line about detection "at any depth" is also unchanged here. The reordered type dispatch is neutral because the old predicate self-guarded with `!node.isObject()`, and the dropped `isTextual()` check is implied by `String.equals`.
Whether a reference nested inside another reference's metadata should be resolved is a fair question, but it's pre-existing behaviour and out of scope for a Jackson-compatibility fix.

[the remark this answers is not in the document]
Comment (mustafadagher) [reply testing/camunda-process-test-java/src/main/java/io/camunda/process/test/impl/judge/DocumentReferenceResolver.java:121]: Also, the current behavior is defensible; metadata describes the outer document, so recursing would pull in documents that were never variable content.

**Review by saig0 — APPROVED** [review]
@mustafadagher nice! I like the result. I looks clean and simple. 🎉
